pkg/client/limiter.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"net/url"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const maxConcurrentRequests = 100
+
+var inflightRequests = metrics.NewGauge(
+	&metrics.GaugeOpts{
+		Namespace: "prometheus_adapter",
+		Subsystem: "prometheus_client",
+		Name:      "inflight_requests",
+		Help:      "Number of inflight requests to the Prometheus service",
+	},
+)
+
+func init() {
+	legacyregistry.MustRegister(inflightRequests)
```

Comment:
note: we will have to change the registry once https://github.com/kubernetes-sigs/prometheus-adapter/pull/599 is merged and rebased

Reply:
On second thought, I've added a max_requests metric so we can even alert on the prometheus adapter being throttled (e.g. inflight/max == 1 for a significant time).